_posts/2020-09-01-processing-java-stream.md

```diff
@@ -0,0 +1,123 @@
+                .map(e -> {
+                    System.out.println("Stream - map:" + e);
+                    return e.toUpperCase();
+                })
+                .collect(Collectors.toList());
+    }
+```
+Z logów wyświetlonych w konsoli widzimy, że operacje wykonywane są horyzontalnie, to znaczy dla każdego elementu od początku do końca. Dzięki temu, sekwencja skończy się na pierwszym niespełnionym warunku.
+```
+Stream - filter: aab
+Stream - map:aab
+Stream - filter: aab
+Stream - map:aab
+Stream - filter: aac
+Stream - map:aac
+Stream - filter: aac
+Stream - map:aac
+Stream - filter: bbb
+Stream - filter: ccc
+Stream - filter: ddd
+```
+
+# Przetwarzanie równoległe
+Wartościowym mechanizmem, który został dodany do Stream API jest możliwość równoległego przetwarzania strumieni, co daje możliwość wykorzystania większej liczby rdzeni procesora do wykonania zadania.
+Jak widać na załączonym kodzie wystarczy, że skorzystamy z metody **parallelStream** zamiast **stream** i nasze dane zostaną odpowiednio podzielone i przetworzone równolegle z użyciem wielu wątków. Warto zwrócić uwagę na to, że samo dzielenie danych wymaga dodatkowej pracy, dlatego przetwarzanie równoległe mniejszej ilości danych może trwać dłużej niż zrobienie tego samego z wykorzystaniem jednego wątku.
+``` java
+    private static final List<String> LIST = Arrays.asList("aab", "aab", 
+                                                           "aac", "aac", 
+                                                           "bbb", "ccc", "ddd");
+    public static void main(String[] args) {
+        LIST.parallelStream()
+                .filter(e -> {
+                    System.out.println("Stream - filter: " + e);
+                    return e.startsWith("a");
+                })
+                .map(e -> {
+                    System.out.println("Stream - map:" + e);
+                    return e.toUpperCase();
+                })
+                .collect(Collectors.toList());
+    }
+```
+
+# Benchmark
+Wiemy już jak przetwarzać strumienie w bardziej efektywnie. Nasuwa się pytanie jak ich szybkość wypada w porównaniu do pętli, aby to sprawdzić zostały przeprowadzone testy przy użyciu Java Microbenchmark Harness i procesora Intel i7 10700K. Kod zaprezentowany niżej został uruchomiony na kolekcjach liczącej 1, 100, 10 000 oraz 1 000 000 elementów.
+
+
+Jednostką pomiarową jest ilość operacji na sekundę, czyli innymi słowy ile razy udało się wywołać daną metodę w ciągu jednej sekundy.
+Jak widać na wykresie dla 1 elementu, pętla okazała się zdecydowanie szybsza niż strumienie. Test 100 elementów pokazał że strumienie powoli doganiają zwykłą pętlę, jednocześnie strumień równoległy okazał się zdecydowanie najwolniejszym rozwiązaniem. W kolejnych testach dla 10 000 i 1 000 000 elementów, widzimy siłę przetwarzania równoległego. Dodatkowo warto zauważyć, że dla testu największej kolekcji, przetwarzanie w pętli uzyskało delikatnie słabszy wynik niż w jednowątkowym strumieniu.
```

Comment:
Czytam sobie ten artykuł i nagle mi się urwał. Brakuje mi tutaj jakiegoś zakończenia, klamry domykającej. Zgaduję, że chciałeś być bezstronny tylko fachowo przedstawić fakty, jednakże moim zdaniem warto byłoby to podsumować. Coś od Ciebie na ten temat- szukając potwierdzenia moich słów odpaliłem poprzedni wpis blogowy i tam jest to zrobione znakomicie:
https://blog.consdata.tech/2020/08/27/axon-kompleksowe-testowanie-aplikacji.html